p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,198 @@
+import java.io.InputStream;
+import java.net.URI;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Handles files discovery over the FTP protocol.
+ *
+ * @author Edoardo Luppi
+ */
+@Component(role = TransportProtocolHandler.class, hint = "ftp")
+public class FtpTransportProtocolHandler implements TransportProtocolHandler {
+    private final Map<String, FTPClient> clients = new ConcurrentHashMap<>(8);
+
+    @Requirement
+    private Logger logger;
+
+    @Requirement
+    private TransportCacheConfig cacheConfig;
+
+    @Requirement
+    private MavenAuthenticator authenticator;
+
+    @Override
+    public long getLastModified(final URI uri) throws IOException {
+        final FTPClient client = getClient(uri);
+
+        if (!client.hasFeature(FTPCmd.MDTM)) {
+            return -1;
+        }
+
+        final FTPFile file = client.mdtmFile(uri.getPath());
+
+        if (file == null) {
+            throw new IOException("Could not retrieve the last modification timestamp for " + uri);
+        }
+
+        return file.getTimestampInstant().toEpochMilli();
+    }
+
+    @Override
+    public File getFile(final URI uri) throws IOException {
+        final File localFile = getLocalFile(uri);
+
+        if (cacheConfig.isOffline()) {
+            if (!localFile.isFile()) {
+                throw new IOException("Maven is offline and the requested URL " + uri + " does not exist locally");
+            }
+
+            return localFile;
+        }
+
+        final FTPClient client = getClient(uri);
+        final String remotePath = uri.getPath();
+        final FTPFile remoteFile = getRemoteFile(client, remotePath);
+        final boolean isRemoteMissing = remoteFile == null;
+
+        if (localFile.isFile() && (isRemoteMissing || !mustRefresh(localFile, remoteFile))) {
+            return localFile;
+        }
+
+        if (isRemoteMissing) {
+            logger.warn("Could not find file " + remotePath);
+            return null;
+        }
+
+        final File parent = FileUtils.createParentDirectories(localFile);
+        final File tempFile = File.createTempFile("download", ".tmp", parent);
+
+        try (final InputStream is = client.retrieveFileStream(remotePath)) {
+            try (final FileOutputStream os = new FileOutputStream(tempFile)) {
+                IOUtils.copyLarge(is, os);
+            }
+        } catch (final IOException e) {
+            tempFile.delete();
+            throw e;
+        }
+
+        if (!client.completePendingCommand()) {
+            throw new IOException(String.format("Error retrieving file %s. %s", remotePath, client.getReplyString()));
+        }
+
+        if (localFile.isFile()) {
+            FileUtils.forceDelete(localFile);
+        }
+
+        FileUtils.moveFile(tempFile, localFile);
+        localFile.setLastModified(remoteFile.getTimestampInstant().toEpochMilli());
+        return localFile;
+    }
+
+    @PreDestroy
```

Comment:
Is `PreDestroy` actually called? I think Plexus uses `Initlizable` and `Disposable`?

Reply:
@laeubi it is definitely called, at least in Maven 3.8.6+. Tested via debug, and I use it in multiple plugins in our env.

Reply:
Good to know :+1:

Reply:
@laeubi a question: why are we still using the old Plexus annotations and not the Javax ones? Do we need to keep some sort of compatibility?

Reply:
1. As long as it works it doesn't matter and there is no advantage in using "new" annotations
2. the JSR annotation do not work with Java 17 in Maven 3.8.x